examples/arcgis-rest-capabilities.js

```diff
@@ -0,0 +1,171 @@
+
+var map;
+
+var getResolutionsFromConfig = function(config) {
+  var tileInfo = config['tileInfo'];
+  if (tileInfo) {
+    var resolutions = [];
+    for (var i = 0, ii = tileInfo.lods.length; i < ii; ++i) {
+      resolutions.push(tileInfo.lods[i].resolution);
+    }
+    return resolutions;
+  }
+};
+
+var getProjectionFromConfig = function(config) {
+  var epsg = 'EPSG:' + config.spatialReference.wkid;
+  var units = config.units === 'esriMeters' ? 'm' : 'degrees';
+  var projection = ol.proj.get(epsg) ? ol.proj.get(epsg) :
+      new ol.proj.Projection({code: epsg, units: units});
+  return projection;
+};
+
+var createXYZSource = function(config, serviceUrl, attributions, projection,
+    resolutions) {
+  var tileInfo = config['tileInfo'];
+  var tileSize = [
+    tileInfo.width || tileInfo.cols,
+    tileInfo.height || tileInfo.rows
+  ];
+  var tileOrigin = [
+    tileInfo.origin.x,
+    tileInfo.origin.y
+  ];
+  var urls;
+  var suffix = '/tile/{z}/{y}/{x}';
+  if (config.tileServers) {
+    urls = config.tileServers;
+    for (var i = 0, ii = urls.length; i < ii; ++i) {
+      urls[i] += suffix;
+    }
+  } else {
+    urls = [serviceUrl += suffix];
+  }
+  var width = tileSize[0] * resolutions[0];
+  var height = tileSize[1] * resolutions[0];
+  var tileUrlFunction, extent, tileGrid;
+  if (projection.getCode() === 'EPSG:4326') {
+    tileUrlFunction = function(tileCoord) {
+      var url = urls.length === 1 ? urls[0] :
+          urls[Math.floor(Math.random() * (urls.length - 0 + 1)) + 0];
+      return url.replace('{z}', (tileCoord[0] - 1).toString())
+          .replace('{x}', tileCoord[1].toString())
+          .replace('{y}', (-tileCoord[2] - 1).toString());
+    };
+  } else {
+    extent = [
+      tileOrigin[0],
+      tileOrigin[1] - height,
+      tileOrigin[0] + width,
+      tileOrigin[1]
+    ];
+    tileGrid = new ol.tilegrid.TileGrid({
+      origin: tileOrigin,
+      resolutions: resolutions
+    });
+  }
+  return {
+    extent: extent,
+    source: new ol.source.XYZ({
+      attributions: attributions,
+      projection: projection,
+      extent: extent,
+      tileSize: tileSize,
+      tileGrid: tileGrid,
+      tileUrlFunction: tileUrlFunction,
+      urls: urls
+    })
+  };
+};
+
+var createArcGISRestSource = function(attributions, serviceUrl) {
+  return {
+    source: new ol.source.TileArcGISRest({
+      url: serviceUrl,
+      attributions: attributions
+    })
+  };
+};
+
+var createLayerFromConfig = function(config, serviceUrl, projection,
+    resolutions) {
+  var attributions = [
+    new ol.Attribution({
+      html: config['copyrightText']
+    })
+  ];
+  var sourceInfo;
+  if (config.tileInfo) {
+    sourceInfo = createXYZSource(config, serviceUrl, attributions, projection,
+        resolutions);
+  } else {
+    sourceInfo = createArcGISRestSource(attributions, serviceUrl);
+  }
+  return new ol.layer.Tile({
+    source: sourceInfo.source,
+    extent: sourceInfo.extent
```

Comment:
I don't think it's good to restrict the extent on the layer level. Should be done on the tilegrid level instead IMHO.